### PR TITLE
Add credential helper documentation

### PR DESCRIPTION
**Pull request by dekkagaijin**
Follow-up to #25851 

sleep tight pupper

Signed-off-by: Jake Sanders <jsand@google.com>

### CODE REVIEW

**Comment by KingEmet** [comment]
Dat palendrome doe...

File: docs/reference/commandline/login.md [hunk 1]
```diff
@@ -120,3 +122,31 @@ an example of that payload: `https://index.docker.io/v1`.
 
 The `erase` command can write error messages to `STDOUT` that the docker engine
 will show if there was an issue.
+
+## Credential helpers
+
+Credential helpers are similar to the credential store above, but act as the
+designated programs to handle credentials for *specific registries*. The default
+credential store (`credsStore` or the config file itself) will not be used for
+operations concering credentials of the specified registries.
```
Comment (thaJeztah) [inline docs/reference/commandline/login.md]: typo; `concering` -> `concerning`

File: docs/reference/commandline/cli.md [hunk 2]
```diff
@@ -171,7 +185,12 @@ Following is a sample `config.json` file:
       "imagesFormat": "table {{.ID}}\\t{{.Repository}}\\t{{.Tag}}\\t{{.CreatedAt}}",
       "statsFormat": "table {{.Container}}\t{{.CPUPerc}}\t{{.MemUsage}}",
       "serviceInspectFormat": "pretty",
-      "detachKeys": "ctrl-e,e"
+      "detachKeys": "ctrl-e,e",
+      "credsStore": "secretservice",
+      "credHelpers": {
+        "kontainr.ly": "hip-star",
+        "sfo.unicorn.com": "vcbait"
```
Comment (thaJeztah) [inline docs/reference/commandline/cli.md]: for documentation, perhaps use the `example.*` domain's, which are intended for that (the domains here may actually exist at some point). e.g., `my-registry.example.com`

File: docs/reference/commandline/cli.md [hunk 3]
```diff
@@ -142,6 +142,20 @@ property is not set, the client falls back to the default table
 format. For a list of supported formatting directives, see
 [**Formatting** section in the `docker stats` documentation](stats.md)
 
+The property `credsStore` specifies an external binary to serve as the default
+credential store. When this property is set, `docker login` will attempt to
+store credentials in the binary specified by `docker-credential-<value>` which
+is visible on $PATH. If this property is not set, credentials will be stored
+in the `auths` property of the config. For more information, see the
+[**Credentials store** section in the `docker login` documentation](login.md)
```
Comment (thaJeztah) [inline docs/reference/commandline/cli.md]: can you add an anchor tag here? `login.md#credentials-store`

File: docs/reference/commandline/cli.md [hunk 4]
```diff
@@ -142,6 +142,20 @@ property is not set, the client falls back to the default table
...
+
+The property `credHelpers` specifies a set of credential helpers to use
+preferentially over `credsStore` or `auths` when storing and retrieving
+credentials for specific registries. If this property is set, the binary
+`docker-credential-<value>` will be used when storing or retrieving credentials
+for a specific registry. For more information, see the
+[**Credential helpers** section in the `docker login` documentation](login.md)
```
Comment (thaJeztah) [inline docs/reference/commandline/cli.md]: same here; `login.md#credential-helpers`

File: docs/reference/commandline/cli.md [hunk 5]
```diff
@@ -142,6 +142,20 @@ property is not set, the client falls back to the default table
 format. For a list of supported formatting directives, see
 [**Formatting** section in the `docker stats` documentation](stats.md)
 
+The property `credsStore` specifies an external binary to serve as the default
+credential store. When this property is set, `docker login` will attempt to
+store credentials in the binary specified by `docker-credential-<value>` which
+is visible on $PATH. If this property is not set, credentials will be stored
```
Comment (thaJeztah) [inline docs/reference/commandline/cli.md]: Can you put back-tics around `$PATH`?

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Thanks! I left some notes/suggestions

File: docs/reference/commandline/login.md (hunk 1 above)
Comment (KingEmet) [inline docs/reference/commandline/login.md]: done

File: docs/reference/commandline/cli.md (hunk 2 above)
Comment (KingEmet) [inline docs/reference/commandline/cli.md]: done

File: docs/reference/commandline/login.md [hunk 6]
```diff
@@ -43,7 +43,7 @@ You can log into any public or private repository for which you have
 credentials.  When you log in, the command stores encoded credentials in
 `$HOME/.docker/config.json` on Linux or `%USERPROFILE%/.docker/config.json` on Windows.
 
-## Credentials store
+## <a name="credentials-store">Credentials store</a>
```
Comment (vdemeester) [inline docs/reference/commandline/login.md]: Hum, are thous `<a …` required ? first time I see those in our ref doc.

**Review by vdemeester — CHANGES_REQUESTED** [review]

File: docs/reference/commandline/login.md (hunk 6 above)
Comment (thaJeztah) [inline docs/reference/commandline/login.md]: No, it's not needed, it's automatically generated when converting the Markdown to HTML (including when reading on GitHub)

File: docs/reference/commandline/login.md [hunk 7]
```diff
@@ -144,9 +144,9 @@ For example:
 ```json
 {
   "credHelpers": {
-    "registry.com": "registryhelper",
-    "kontainr.ly": "hip-star",
-    "sfo.unicorn.com": "vcbait"
+    "example.registry.com": "registryhelper",
+    "example.awesomereg.io": "hip-star",
```
Comment (thaJeztah) [inline docs/reference/commandline/login.md]: oh, I meant to use `example.com` or `example.org` as domain name. I know it _should_ be clear to users, but we've had situations where people used the literal examples, and the first domain is actually active (second one is available to buy if you're interested :smile:);

```
Domain Name: REGISTRY.COM
Registrar: GODADDY.COM, LLC
...
Expiration Date: 18-nov-2018
```


perhaps `awesomereg.example.org` ?

File: docs/reference/commandline/cli.md [hunk 8]
```diff
@@ -171,7 +185,12 @@ Following is a sample `config.json` file:
       "imagesFormat": "table {{.ID}}\\t{{.Repository}}\\t{{.Tag}}\\t{{.CreatedAt}}",
       "statsFormat": "table {{.Container}}\t{{.CPUPerc}}\t{{.MemUsage}}",
       "serviceInspectFormat": "pretty",
-      "detachKeys": "ctrl-e,e"
+      "detachKeys": "ctrl-e,e",
+      "credsStore": "secretservice",
+      "credHelpers": {
+        "example.awesomereg.io": "hip-star",
```
Comment (thaJeztah) [inline docs/reference/commandline/cli.md]: Same here (domain name)

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Thanks @KingEmet, I left some comments.

Also, can you squash your commits? No need to preserve history for these review comments

**Comment by KingEmet** [comment]
done?

**Review by vdemeester — APPROVED** [review]
LGTM 🐸

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!!

**Pull request merged**